File: mobile/lib/models/reservation.dart
class ReservationRequest {
  final int idClient;
  final String email;
  final String telephone;
  final DateTime dateReservation;
  final int nombrePersonnes;
  final List<int> platIds;
  final String? commentaire;

  ReservationRequest({
    required this.idClient,
    required this.email,
    required this.telephone,
    required this.dateReservation,
    required this.nombrePersonnes,
    required this.platIds,
    this.commentaire,
  });

  Map<String, dynamic> toJson() {
    return {
      'idClient': idClient,
      'email': email,
      'telephone': telephone,
      'dateReservation': dateReservation.toIso8601String(),
      'nombrePersonnes': nombrePersonnes,
      'platIds': platIds, 
      if (commentaire != null && commentaire!.isNotEmpty)
        'commentaire': commentaire,
    };
  }
}

class ReservationResponse {
  final int id;
  final int? userId;  // Changed from idClient and made nullable
  final String email;
  final String telephone;
  final DateTime dateReservation;
  final int nombrePersonnes;
  final String statut;
  final String? commentaire;
  final DateTime? dateCreation;  // Added field
  final List<dynamic>? plats;    // Added field for dishes

  ReservationResponse({
    required this.id,
    this.userId,  // Made optional
    required this.email,
    required this.telephone,
    required this.dateReservation,
    required this.nombrePersonnes,
    required this.statut,
    this.commentaire,
    this.dateCreation,
    this.plats,
  });

  // Backward compatibility getter
  int? get idClient => userId;

  factory ReservationResponse.fromJson(Map<String, dynamic> json) {
    return ReservationResponse(
      id: json['id'],
      userId: json['userId'] ?? json['idClient'],  // Support both field names
      email: json['email'],
      telephone: json['telephone'],
      dateReservation: DateTime.parse(json['dateReservation']),
      nombrePersonnes: json['nombrePersonnes'],
      statut: json['statut'],
      commentaire: json['commentaire'],
      dateCreation: json['dateCreation'] != null
          ? DateTime.parse(json['dateCreation'])
          : null,
      plats: json['plats'] as List<dynamic>?,
    );
  }
}

class UpdateReservationRequest {
  final String? email;
  final String? telephone;
  final DateTime? dateReservation;
  final int? nombrePersonnes;
  final List<int>? platIds;
  final String? commentaire;

  UpdateReservationRequest({
    this.email,
    this.telephone,
    this.dateReservation,
    this.nombrePersonnes,
    this.platIds,
    this.commentaire,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    if (email != null) data['email'] = email;
    if (telephone != null) data['telephone'] = telephone;
    if (dateReservation != null) {
      data['dateReservation'] = dateReservation!.toIso8601String();
    }
    if (nombrePersonnes != null) data['nombrePersonnes'] = nombrePersonnes;
    if (platIds != null) data['platIds'] = platIds;
    if (commentaire != null) data['commentaire'] = commentaire;
    return data;
  }
}

class AvailabilityRequest {
  final DateTime dateReservation;
  final int nombrePersonnes;

  AvailabilityRequest({
    required this.dateReservation,
    required this.nombrePersonnes,
  });

  Map<String, dynamic> toJson() {
    return {
      'dateReservation': dateReservation.toIso8601String(),
      'nombrePersonnes': nombrePersonnes,
    };
  }
}

class AvailabilityResponse {
  final DateTime dateReservation;
  final int totalCapacity;
  final int reservedSeats;
  final int availableSeats;
  final bool available;

  AvailabilityResponse({
    required this.dateReservation,
    required this.totalCapacity,
    required this.reservedSeats,
    required this.availableSeats,
    required this.available,
  });

  factory AvailabilityResponse.fromJson(Map<String, dynamic> json) {
    return AvailabilityResponse(
      dateReservation: DateTime.parse(json['dateReservation']),
      totalCapacity: json['totalCapacity'],
      reservedSeats: json['reservedSeats'],
      availableSeats: json['availableSeats'],
      available: json['available'],
    );
  }
}

File: api_fastapi/app/schemas/reservation.py
from pydantic import BaseModel, EmailStr, Field, field_validator
from datetime import datetime
from typing import Optional
from app.models.reservation import StatutReservation
from app.schemas.plat import PlatDTO


class ReservationRequest(BaseModel):
    """Request schema for creating a reservation"""
    idClient: Optional[int] = None
    email: EmailStr
    telephone: str
    dateReservation: datetime
    nombrePersonnes: int = Field(..., ge=1, le=20)
    platIds: list[int] = []
    commentaire: Optional[str] = None

    @field_validator('dateReservation')
    @classmethod
    def validate_future_date(cls, v):
        """Ensure reservation date is in the future"""
        if v < datetime.now():
            raise ValueError("La date de réservation doit être dans le futur")
        return v


class UpdateReservationRequest(BaseModel):
    """Request schema for updating a reservation"""
    email: Optional[EmailStr] = None
    telephone: Optional[str] = None
    dateReservation: Optional[datetime] = None
    nombrePersonnes: Optional[int] = Field(None, ge=1, le=20)
    platIds: Optional[list[int]] = None
    commentaire: Optional[str] = None


class ReservationResponse(BaseModel):
    """Response schema for reservation"""
    id: int
    userId: Optional[int] = None
    email: str
    telephone: str
    dateReservation: datetime
    nombrePersonnes: int
    statut: StatutReservation
    commentaire: Optional[str] = None
    dateCreation: datetime
    plats: list[PlatDTO] = []

    class Config:
        from_attributes = True

    @classmethod
    def from_orm(cls, obj):
        """Convert ORM object to DTO"""
        return cls(
            id=obj.id,
            userId=obj.user_id,
            email=obj.email,
            telephone=obj.telephone,
            dateReservation=obj.date_reservation,
            nombrePersonnes=obj.nombre_personnes,
            statut=obj.statut,
            commentaire=obj.commentaire,
            dateCreation=obj.date_creation,
            plats=[PlatDTO.from_orm(plat) for plat in obj.plats],
        )


class AvailabilityRequest(BaseModel):
    """Request schema for checking availability"""
    dateReservation: datetime
    nombrePersonnes: int = Field(..., ge=1, le=20)


class AvailabilityResponse(BaseModel):
    """Response schema for availability check"""
    dateReservation: datetime
    totalCapacity: int
    reservedSeats: int
    availableSeats: int
    available: bool

    class Config:
        from_attributes = True
